compiler/semstmts.nim

```diff
@@ -445,55 +449,114 @@ proc setVarType(c: PContext; v: PSym, typ: PType) =
-    result.add a[2]
-    result.info = a.info
-    let ret = newNodeI(nkStmtList, a.info)
-    ret.add result
-    result = semExprNoType(c, ret)
+    const
+      namePos = 0
+      pragmaPos = 1
+    let pragmas = b[pragmaPos]
+    for i in 0 ..< pragmas.len:
+      # Mirrored with semProcAnnotation
+
+      let it = pragmas[i]
+      let key = if it.kind in nkPragmaCallKinds and it.len >= 1: it[0] else: it
+      
+      when false:
+        let lhs = b[0]
+        let clash = strTableGet(c.currentScope.symbols, lhs.ident)
+        if clash != nil:
+          # refs https://github.com/nim-lang/Nim/issues/8275
+          wrongRedefinition(c, lhs.info, lhs.ident.s, clash.info)
+
+      if whichPragma(it) != wInvalid or key.kind notin nkIdentKinds:
+        # Not a custom pragma
+        continue
+      else:
+        let ident = considerQuotedIdent(c, key)
+        if strTableGet(c.userPragmas, ident) != nil:
+          continue # User defined pragma
+        else:
+          var amb = false
+          let sym = searchInScopes(c, ident, amb)
+          if sym != nil and sfCustomPragma in sym.flags:
+            continue # User custom pragma
+
+      # we transform ``proc p {.m, rest.}`` into ``m(do: proc p {.rest.})`` and
+      # let the semantic checker deal with it:
+      var x = newNodeI(nkCall, key.info)
+      x.add(key)
+
+      if it.kind in nkPragmaCallKinds and it.len > 1:
+        # pass pragma arguments to the macro too:
+        for i in 1..<it.len:
+          x.add(it[i])
+
+      # Drop the pragma from the list, this prevents getting caught in endless
+      # recursion when the nkCall is semanticized
+      let oldExpr = a[lhsPos]
+      let newPragmas = copyExcept(pragmas, i)
+      if newPragmas.kind != nkEmpty and newPragmas.len == 0:
+        a[lhsPos] = oldExpr[namePos]
+      else:
+        a[lhsPos] = copyNode(oldExpr)
+        a[lhsPos].add(oldExpr[namePos])
+        a[lhsPos].add(newPragmas)
+
+      var unarySection = newNodeI(n.kind, a.info)
+      unarySection.add(a)
+      x.add(unarySection)
+
+      # recursion assures that this works for multiple macro annotations too:
+      var r = semOverloadedCall(c, x, x, {skMacro, skTemplate}, {efNoUndeclared})
+      if r == nil:
+        # Restore the old list of pragmas since we couldn't process this
+        a[lhsPos] = oldExpr
+        # No matching macro was found but there's always the possibility this may
+        # be a .pragma. template instead
+        continue
+
+      doAssert r[0].kind == nkSym
+      let m = r[0].sym
+      case m.kind
+      of skMacro: result = semMacroExpr(c, r, r, m, {})
+      of skTemplate: result = semTemplateExpr(c, r, m, {})
+      else:
+        a[lhsPos] = oldExpr
+        continue
+
+      doAssert result != nil
+
+      # since a macro pragma can set pragmas, we process these here again.
+      # This is required for SqueakNim-like export pragmas.
+      if result.kind in {nkVarSection, nkLetSection, nkConstSection}:
```

Comment:
I don't understand the comment here and how this part is required here and in `semProcAnnotation`. Shouldn't the recursion solve it?